File: scripts/wallet_manager.py
#!/usr/bin/env python3
"""
TrackerClaw — Wallet Manager
Manages wallet labels + addresses in data/wallets.json.

Usage:
    python wallet_manager.py list
    python wallet_manager.py add <label> <address>
    python wallet_manager.py remove <label>
    python wallet_manager.py rename <old_label> <new_label>
"""

import json
import os
import sys

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
PROJECT_DIR = os.path.dirname(SCRIPT_DIR)
WALLETS_FILE = os.path.join(PROJECT_DIR, "data", "wallets.json")
RAW_WALLETS_FILE = os.path.join(PROJECT_DIR, "wallets")


def load_wallets() -> list:
    """Load wallets from JSON file, seeding from raw wallets file if needed."""
    os.makedirs(os.path.dirname(WALLETS_FILE), exist_ok=True)

    if os.path.exists(WALLETS_FILE):
        with open(WALLETS_FILE, "r") as f:
            return json.load(f)

    # Seed from raw wallets file if wallets.json doesn't exist yet
    wallets = []
    if os.path.exists(RAW_WALLETS_FILE):
        with open(RAW_WALLETS_FILE, "r") as f:
            for i, line in enumerate(f):
                addr = line.strip()
                if addr:
                    wallets.append({
                        "label": f"Wallet{i + 1}" if i > 0 else "Main",
                        "address": addr
                    })
    save_wallets(wallets)
    return wallets


def save_wallets(wallets: list):
    """Save wallets to JSON file."""
    os.makedirs(os.path.dirname(WALLETS_FILE), exist_ok=True)
    with open(WALLETS_FILE, "w") as f:
        json.dump(wallets, f, indent=2)


def cmd_list():
    """List all tracked wallets."""
    wallets = load_wallets()
    if not wallets:
        print("No wallets tracked yet. Use 'add <label> <address>' to add one.")
        return

    print(f"📋 **Tracked Wallets** ({len(wallets)})\n")
    for w in wallets:
        short = w["address"][:6] + "..." + w["address"][-4:]
        print(f"• **{w['label']}** — `{short}` ({w['address']})")


def cmd_add(label: str, address: str):
    """Add a new wallet."""
    wallets = load_wallets()

    # Check for duplicate label
    for w in wallets:
        if w["label"].lower() == label.lower():
            print(f"❌ Wallet with label '{label}' already exists.")
            return

    # Check for duplicate address
    for w in wallets:
        if w["address"] == address:
            print(f"❌ Address already tracked under label '{w['label']}'.")
            return

    # Validate address (basic Solana address check: base58, 32-44 chars)
    if not (32 <= len(address) <= 44):
        print(f"❌ Invalid Solana address (expected 32-44 characters, got {len(address)}).")
        return

    wallets.append({"label": label, "address": address})
    save_wallets(wallets)
    short = address[:6] + "..." + address[-4:]
    print(f"✅ Added wallet **{label}** — `{short}`")


def cmd_remove(label: str):
    """Remove a wallet by label."""
    wallets = load_wallets()
    original_len = len(wallets)
    wallets = [w for w in wallets if w["label"].lower() != label.lower()]

    if len(wallets) == original_len:
        print(f"❌ No wallet found with label '{label}'.")
        return

    save_wallets(wallets)
    print(f"✅ Removed wallet **{label}**")


def cmd_rename(old_label: str, new_label: str):
    """Rename a wallet."""
    wallets = load_wallets()

    # Check new label doesn't conflict
    for w in wallets:
        if w["label"].lower() == new_label.lower():
            print(f"❌ Label '{new_label}' is already in use.")
            return

    found = False
    for w in wallets:
        if w["label"].lower() == old_label.lower():
            w["label"] = new_label
            found = True
            break

    if not found:
        print(f"❌ No wallet found with label '{old_label}'.")
        return

    save_wallets(wallets)
    print(f"✅ Renamed **{old_label}** → **{new_label}**")


def main():
    if len(sys.argv) < 2:
        print(__doc__)
        sys.exit(1)

    command = sys.argv[1].lower()

    if command == "list":
        cmd_list()
    elif command == "add":
        if len(sys.argv) < 4:
            print("Usage: python wallet_manager.py add <label> <address>")
            sys.exit(1)
        cmd_add(sys.argv[2], sys.argv[3])
    elif command == "remove":
        if len(sys.argv) < 3:
            print("Usage: python wallet_manager.py remove <label>")
            sys.exit(1)
        cmd_remove(sys.argv[2])
    elif command == "rename":
        if len(sys.argv) < 4:
            print("Usage: python wallet_manager.py rename <old_label> <new_label>")
            sys.exit(1)
        cmd_rename(sys.argv[2], sys.argv[3])
    else:
        print(f"Unknown command: {command}")
        print(__doc__)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: scriptsTS/wallet_manager.ts
#!/usr/bin/env node

import fs from "node:fs/promises";

import { DATA_DIR, RAW_WALLETS_FILE, WALLETS_FILE } from "./lib/paths.js";
import { readJsonFile, shortAddress, writeJsonFile } from "./lib/utils.js";
import type { WalletRecord } from "./lib/types.js";

export async function loadWallets(): Promise<WalletRecord[]> {
  await fs.mkdir(DATA_DIR, { recursive: true });

  try {
    await fs.access(WALLETS_FILE);
    return await readJsonFile<WalletRecord[]>(WALLETS_FILE, []);
  } catch {
    const wallets: WalletRecord[] = [];
    try {
      const raw = await fs.readFile(RAW_WALLETS_FILE, "utf8");
      raw.split(/\r?\n/).forEach((line, index) => {
        const address = line.trim();
        if (!address) {
          return;
        }
        wallets.push({
          label: index === 0 ? "Main" : `Wallet${index + 1}`,
          address,
        });
      });
    } catch {
      return [];
    }

    await saveWallets(wallets);
    return wallets;
  }
}

export async function saveWallets(wallets: WalletRecord[]): Promise<void> {
  await fs.mkdir(DATA_DIR, { recursive: true });
  await writeJsonFile(WALLETS_FILE, wallets);
}

export async function loadWalletsByLabel(walletFilter?: string): Promise<WalletRecord[]> {
  const wallets = await loadWallets();
  if (!walletFilter) {
    return wallets;
  }
  const filtered = wallets.filter((wallet) => wallet.label.toLowerCase() === walletFilter.toLowerCase());
  if (!filtered.length) {
    throw new Error(`No wallet found with label '${walletFilter}'.`);
  }
  return filtered;
}

export async function cmdList(): Promise<void> {
  const wallets = await loadWallets();
  if (!wallets.length) {
    console.log("No wallets tracked yet. Use 'add <label> <address>' to add one.");
    return;
  }

  console.log(`Tracked Wallets (${wallets.length})\n`);
  for (const wallet of wallets) {
    console.log(`- ${wallet.label} - ${shortAddress(wallet.address)} (${wallet.address})`);
  }
}

export async function cmdAdd(label: string, address: string): Promise<void> {
  const wallets = await loadWallets();

  if (wallets.some((wallet) => wallet.label.toLowerCase() === label.toLowerCase())) {
    throw new Error(`Wallet with label '${label}' already exists.`);
  }
  if (wallets.some((wallet) => wallet.address === address)) {
    const existing = wallets.find((wallet) => wallet.address === address);
    throw new Error(`Address already tracked under label '${existing?.label}'.`);
  }
  if (address.length < 32 || address.length > 44) {
    throw new Error(`Invalid Solana address length: ${address.length}.`);
  }

  wallets.push({ label, address });
  await saveWallets(wallets);
  console.log(`Added wallet ${label} - ${shortAddress(address)}`);
}

export async function cmdRemove(label: string): Promise<void> {
  const wallets = await loadWallets();
  const filtered = wallets.filter((wallet) => wallet.label.toLowerCase() !== label.toLowerCase());
  if (filtered.length === wallets.length) {
    throw new Error(`No wallet found with label '${label}'.`);
  }
  await saveWallets(filtered);
  console.log(`Removed wallet ${label}`);
}

export async function cmdRename(oldLabel: string, newLabel: string): Promise<void> {
  const wallets = await loadWallets();
  if (wallets.some((wallet) => wallet.label.toLowerCase() === newLabel.toLowerCase())) {
    throw new Error(`Label '${newLabel}' is already in use.`);
  }

  let found = false;
  for (const wallet of wallets) {
    if (wallet.label.toLowerCase() === oldLabel.toLowerCase()) {
      wallet.label = newLabel;
      found = true;
      break;
    }
  }
  if (!found) {
    throw new Error(`No wallet found with label '${oldLabel}'.`);
  }

  await saveWallets(wallets);
  console.log(`Renamed ${oldLabel} -> ${newLabel}`);
}

async function main(): Promise<void> {
  const [, , command, ...args] = process.argv;
  if (!command) {
    console.error("Usage: tsx scriptsTS/wallet_manager.ts list|add|remove|rename");
    process.exit(1);
  }

  try {
    if (command === "list") {
      await cmdList();
      return;
    }
    if (command === "add") {
      if (args.length < 2) {
        throw new Error("Usage: add <label> <address>");
      }
      await cmdAdd(args[0], args[1]);
      return;
    }
    if (command === "remove") {
      if (!args[0]) {
        throw new Error("Usage: remove <label>");
      }
      await cmdRemove(args[0]);
      return;
    }
    if (command === "rename") {
      if (args.length < 2) {
        throw new Error("Usage: rename <old_label> <new_label>");
      }
      await cmdRename(args[0], args[1]);
      return;
    }

    throw new Error(`Unknown command: ${command}`);
  } catch (error) {
    console.error(error instanceof Error ? error.message : String(error));
    process.exit(1);
  }
}

if (process.argv[1] && process.argv[1].endsWith("wallet_manager.ts")) {
  void main();
}
